libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -43,29 +66,52 @@ public AzureBlobStorage(CloudStorageAccount storageAccount, string containerName
               });
         }
 
-        protected string GetKey(string key)
+        /// <summary>
+        /// Get a blob name validated representation of an entity
+        /// </summary>
+        /// <param name="key">The key used to identify the entity</param>
+        /// <returns></returns>
+        private static string GetBlobName(string key)
         {
-            return HttpUtility.UrlEncode(key);
+            if (string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+
+            var blobName = HttpUtility.UrlEncode(key);
+            NameValidator.ValidateBlobName(blobName);
+            return blobName;
         }
 
+        /// <summary>
+        /// Deletes entity blobs from the configured container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
         public async Task Delete(string[] keys)
         {
-            foreach (var key in keys.Select(k => GetKey(k)))
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            foreach (var blobName in keys.Select(GetBlobName))
             {
-                var blobReference = this.Container.Value.GetBlobReference(key);
+                var blobReference = this.Container.Value.GetBlobReference(blobName);
                 await blobReference.DeleteAsync(
                     DeleteSnapshotsOption.None, new AccessCondition { IfMatchETag = "*" }, new BlobRequestOptions(), new OperationContext()).ConfigureAwait(false);
             }
         }
 
+        /// <summary>
+        /// Retrieve entities from the configured blob container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
         public async Task<StoreItems> Read(string[] keys)
         {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
             var storeItems = new StoreItems();
-            foreach (string key in keys.Select(k => GetKey(k)))
+            foreach (string blobName in keys.Select(GetBlobName))
             {
                 using (var memoryStream = new MemoryStream())
                 {
-                    var blobReference = this.Container.Value.GetBlobReference(key);
+                    var blobReference = this.Container.Value.GetBlobReference(blobName);
                     await blobReference.DownloadToStreamAsync(memoryStream).ConfigureAwait(false);

```

Comment:
Am I missing that these are happening in parallel? It doesn't seem like it, and it seems like they should be run in parallel...